File: ios/Halogen/Components/ToastCenter.swift
import Observation
import SwiftUI

/// Lightweight global toast queue — the native counterpart of the web's
/// ui-toast surface for actions that have no inline error/result slot
/// (background mutations, toolbar triggers, admin actions).
@MainActor
@Observable
final class ToastCenter {
    struct Toast: Identifiable, Equatable {
        let id = UUID()
        let message: String
        let isError: Bool
    }

    static let shared = ToastCenter()

    private(set) var toasts: [Toast] = []

    func success(_ message: String) { push(Toast(message: message, isError: false)) }
    func error(_ message: String) { push(Toast(message: message, isError: true)) }

    func dismiss(_ id: UUID) {
        toasts.removeAll { $0.id == id }
    }

    private func push(_ toast: Toast) {
        // Dedupe + cap: a burst (e.g. a drained batch dead-lettering) must
        // not stack the screen full of identical capsules.
        guard !toasts.contains(where: { $0.message == toast.message }) else { return }
        if toasts.count >= 4 { toasts.removeFirst() }
        toasts.append(toast)
        Task { [id = toast.id] in
            try? await Task.sleep(for: .seconds(4))
            dismiss(id)
        }
    }
}

/// Bottom-stacked toast overlay; attach once at the root (RootView).
struct ToastOverlay: ViewModifier {
    @State private var center = ToastCenter.shared

    func body(content: Content) -> some View {
        content.overlay(alignment: .bottom) {
            VStack(spacing: 8) {
                ForEach(center.toasts) { toast in
                    Text(toast.message)
                        .font(.footnote)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 14)
                        .padding(.vertical, 10)
                        .background(
                            Capsule().fill(
                                toast.isError
                                    ? Color.red.opacity(0.92)
                                    : Color(.secondarySystemBackground))
                        )
                        .foregroundStyle(toast.isError ? Color.white : Color.primary)
                        .shadow(radius: 4)
                        .onTapGesture { center.dismiss(toast.id) }
                }
            }
            .padding(.horizontal, 24)
            .padding(.bottom, 72)
            .animation(.easeInOut(duration: 0.2), value: center.toasts)
        }
    }
}

extension View {
    func toastOverlay() -> some View { modifier(ToastOverlay()) }
}

File: android/app/src/main/kotlin/org/fgsec/halogen/components/ToastCenter.kt
package org.fgsec.halogen.components

import androidx.compose.animation.animateContentSize
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/// Lightweight global toast queue — the native counterpart of the web's
/// ui-toast surface for actions that have no inline error/result slot
/// (background mutations, toolbar triggers, admin actions).
object ToastCenter {
    data class Toast(val id: Long, val message: String, val isError: Boolean)

    // All mutations hop to Main — callable from any thread (iOS @MainActor parity).
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var nextId = 0L

    var toasts by mutableStateOf(listOf<Toast>())
        private set

    fun success(message: String) = push(message, isError = false)
    fun error(message: String) = push(message, isError = true)

    fun dismiss(id: Long) {
        scope.launch { toasts = toasts.filterNot { it.id == id } }
    }

    private fun push(message: String, isError: Boolean) {
        scope.launch {
            // Dedupe + cap: a burst (e.g. a drained batch dead-lettering) must
            // not stack the screen full of identical capsules.
            if (toasts.any { it.message == message }) return@launch
            val toast = Toast(id = ++nextId, message = message, isError = isError)
            var next = toasts
            if (next.size >= 4) next = next.drop(1)
            toasts = next + toast
            scope.launch {
                delay(4_000)
                toasts = toasts.filterNot { it.id == toast.id }
            }
        }
    }
}

/// Bottom-stacked toast overlay; place once at the root, aligned BottomCenter
/// over the app content (iOS `.toastOverlay()`).
@Composable
fun ToastHost(modifier: Modifier = Modifier) {
    Column(
        modifier
            .navigationBarsPadding()
            .padding(horizontal = 24.dp)
            .padding(bottom = 72.dp)
            .animateContentSize(tween(200)),
        verticalArrangement = Arrangement.spacedBy(8.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        for (toast in ToastCenter.toasts) {
            key(toast.id) {
                Surface(
                    onClick = { ToastCenter.dismiss(toast.id) },
                    shape = RoundedCornerShape(50),
                    color = if (toast.isError)
                        MaterialTheme.colorScheme.error.copy(alpha = 0.92f)
                    else
                        halogenExtras.toastBackground,
                    contentColor = if (toast.isError)
                        MaterialTheme.colorScheme.onError
                    else
                        MaterialTheme.colorScheme.onSurface,
                    shadowElevation = 4.dp,
                ) {
                    Text(
                        toast.message,
                        style = MaterialTheme.typography.bodySmall,
                        textAlign = TextAlign.Center,
                        modifier = Modifier.padding(horizontal = 14.dp, vertical = 10.dp),
                    )
                }
            }
        }
    }
}
